{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66b72119",
   "metadata": {},
   "source": [
    "# DB에서 필요한 정보 추출"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a429affb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: psycopg2 in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (2.9.9)\r\n"
     ]
    }
   ],
   "source": [
    "!pip3 install psycopg2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0e58d412",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: geopy in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (2.4.1)\r\n",
      "Requirement already satisfied: geographiclib<3,>=1.52 in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (from geopy) (2.0)\r\n"
     ]
    }
   ],
   "source": [
    "!pip3 install geopy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e3d97beb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "현재 위치의 GPS 좌표: 37.5660 126.9784\n"
     ]
    }
   ],
   "source": [
    "#### 현재 pc위치의 위도/경도 잡기 ####\n",
    "import requests\n",
    "\n",
    "response = requests.get('https://ipinfo.io')\n",
    "data = response.json()\n",
    "\n",
    "location = data.get('loc')\n",
    "if location:\n",
    "    latitude, longitude = location.split(',')\n",
    "    print(\"현재 위치의 GPS 좌표:\", latitude, longitude)\n",
    "else:\n",
    "    print(\"위치 정보를 가져올 수 없습니다.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "55480626",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-20 14:09:56.174645\n"
     ]
    }
   ],
   "source": [
    "#### 현재 날짜시간 불러오기 #### \n",
    "from datetime import datetime\n",
    "current_datetime = datetime.now()\n",
    "print(current_datetime)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca39b5cc",
   "metadata": {},
   "source": [
    "# PostgreSQL DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "148c69e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "from geopy.geocoders import Nominatim\n",
    "from decimal import Decimal\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "# 1. PostgreSQL 데이터베이스 연결\n",
    "db_params = {'dbname': 'test',\n",
    "             'user': 'postgres',\n",
    "             'password': 'postgres',\n",
    "             'host': 'localhost',\n",
    "             'port': '5432'\n",
    "            }\n",
    "\n",
    "conn = psycopg2.connect(**db_params)\n",
    "cursor = conn.cursor()\n",
    "\n",
    "\n",
    "# 2. 현재 pc가 위치한 gps좌표 수집\n",
    "response = requests.get('https://ipinfo.io')\n",
    "data = response.json()\n",
    "\n",
    "location = data.get('loc')\n",
    "if location:\n",
    "    lat, long = location.split(',')\n",
    "else:\n",
    "    print(\"위치 정보를 가져올 수 없습니다.\")\n",
    "\n",
    "    \n",
    "# 3. 현재 날짜시간정보 수집\n",
    "current_datetime = datetime.now()\n",
    "\n",
    "\n",
    "# 4. 'gps_data' 테이블에 위에서 얻어낸 데이터들을 삽입\n",
    "query = 'INSERT INTO gps_data (latitude, longitude, datetime) VALUES (%s, %s, %s)'\n",
    "values = (latitude, longitude, current_datetime)\n",
    "\n",
    "\n",
    "cursor.execute(query, values)\n",
    "conn.commit()\n",
    "\n",
    "cursor.close()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "363b5c56",
   "metadata": {},
   "source": [
    "# aws S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "96298b87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting boto3\n",
      "  Downloading boto3-1.34.23-py3-none-any.whl.metadata (6.6 kB)\n",
      "Collecting botocore<1.35.0,>=1.34.23 (from boto3)\n",
      "  Downloading botocore-1.34.23-py3-none-any.whl.metadata (5.6 kB)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (from boto3) (0.10.0)\n",
      "Collecting s3transfer<0.11.0,>=0.10.0 (from boto3)\n",
      "  Downloading s3transfer-0.10.0-py3-none-any.whl.metadata (1.7 kB)\n",
      "Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (from botocore<1.35.0,>=1.34.23->boto3) (2.8.2)\n",
      "Requirement already satisfied: urllib3<2.1,>=1.25.4 in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (from botocore<1.35.0,>=1.34.23->boto3) (1.26.16)\n",
      "Requirement already satisfied: six>=1.5 in /Users/juhyeon/anaconda3/lib/python3.11/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.35.0,>=1.34.23->boto3) (1.16.0)\n",
      "Downloading boto3-1.34.23-py3-none-any.whl (139 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m139.3/139.3 kB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading botocore-1.34.23-py3-none-any.whl (11.9 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.9/11.9 MB\u001b[0m \u001b[31m6.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading s3transfer-0.10.0-py3-none-any.whl (82 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m82.1/82.1 kB\u001b[0m \u001b[31m6.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: botocore, s3transfer, boto3\n",
      "  Attempting uninstall: botocore\n",
      "    Found existing installation: botocore 1.29.76\n",
      "    Uninstalling botocore-1.29.76:\n",
      "      Successfully uninstalled botocore-1.29.76\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "aiobotocore 2.5.0 requires botocore<1.29.77,>=1.29.76, but you have botocore 1.34.23 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed boto3-1.34.23 botocore-1.34.23 s3transfer-0.10.0\n"
     ]
    }
   ],
   "source": [
    "!pip3 install boto3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "19ba7ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "\n",
    "def s3_connection():\n",
    "    try:\n",
    "        # s3 클라이언트 생성\n",
    "        s3 = boto3.client(\n",
    "            service_name=\"s3\",\n",
    "            region_name=\"ap-northeast-2\",\n",
    "            aws_access_key_id=\"AKIA43IC3LBBCBPD3VEM\",\n",
    "            aws_secret_access_key=\"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\",\n",
    "        )\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "    else:\n",
    "        print(\"s3 bucket connected!\") \n",
    "        return s3\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "99a49817",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s3 bucket connected!\n"
     ]
    }
   ],
   "source": [
    "s3 = s3_connection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a31a66bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image uploaded to S3: finalprojectimage/test_img_1.jpg\n",
      "Image uploaded to S3: finalprojectimage/test_img_2.jpg\n",
      "Image uploaded to S3: finalprojectimage/test_img_3.jpg\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import boto3\n",
    "\n",
    "# AWS 계정 정보\n",
    "aws_access_key_id = 'AKIA43IC3LBBCBPD3VEM'\n",
    "aws_secret_access_key = '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'\n",
    "aws_region = 'ap-northeast-2'  \n",
    "bucket_name = 'playdataroads'\n",
    "\n",
    "\n",
    "# 업로드할 이미지 (/ROADs/test_img/ 하위에 캡쳐된 이미지가 저장되어 있음 )\n",
    "local_image_directory = '/Users/juhyeon/python-workspace/ROADs/test_img/'\n",
    "image_files = [f for f in os.listdir(local_image_directory) if f.endswith('.jpg')]\n",
    "\n",
    "\n",
    "# AWS S3 클라이언트 생성\n",
    "s3_client = boto3.client('s3', \n",
    "                         aws_access_key_id=aws_access_key_id,\n",
    "                         aws_secret_access_key=aws_secret_access_key, \n",
    "                         region_name=aws_region)\n",
    "\n",
    "\n",
    "# 각 이미지를 S3로 업로드 (파일이름은 똑같이 저장)\n",
    "for image_file in image_files:\n",
    "    local_image_path = os.path.join(local_image_directory, image_file)\n",
    "    s3_object_key = f'finalprojectimage/{image_file}'\n",
    "\n",
    "    # 이미지 업로드\n",
    "    s3_client.upload_file(local_image_path, bucket_name, s3_object_key)\n",
    "\n",
    "    print(f\"Image uploaded to S3: {s3_object_key}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ccfb3c3",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
